pkg/api/plugins.go

```diff
@@ -297,6 +300,154 @@ func (hs *HTTPServer) CollectPluginMetrics(c *models.ReqContext) response.Respon
 	return response.CreateNormalResponse(headers, resp.PrometheusMetrics, http.StatusOK)
 }
 
+// pluginAsset represents a plugin asset.
+type pluginAsset struct {
+	// readSeekCloser is the io.ReadSeekCloser that will be used to read the asset's content.
+	readSeekCloser io.ReadSeekCloser
+
+	// modTime is the latest edit time of the asset.
+	modTime time.Time
+
+	// Path is the relative path of the asset.
+	path string
+}
+
+// Close closes the readSeekCloser associated to the pluginAsset, if any.
+func (a pluginAsset) Close() error {
+	if a.readSeekCloser == nil {
+		return nil
+	}
+	return a.readSeekCloser.Close()
+}
+
+// nopCloserReadSeeker wraps io.ReadSeeker with a nop io.Closer.
+type nopCloserReadSeeker struct {
+	io.ReadSeeker
+}
+
+// Close does nothing.
+func (nopCloserReadSeeker) Close() error {
+	return nil
+}
+
+// errorCoder is an interface that expands error with a Code() and Message() method.
+type errorMessageCoder interface {
+	// Original error.
+	error
+
+	// Message returns a human-readable message, with no sensitive information.
+	Message() string
+
+	// Code returns the error code associated to the error.
+	Code() int
+}
```

Comment:
There must  be already a way to handle error + http status code in the Grafana codebase, but I couldn't find anything. Maybe I didn't look hard enough. Some help on this will be appreciated :grinning:

Reply:
Answer: errutil package